onnx/defs/tensor/defs.cc

```diff
@@ -2470,6 +2470,302 @@ ONNX_OPERATOR_SET_SCHEMA(
+                  grid_h_steps_float = CastLike (grid_h_steps_int, step_h)
+                  grid_h_steps = Mul (grid_h_steps_float, step_h)
+                  grid_h_0 = Add (start_h, grid_h_steps)
+
+                  grid_d_steps_int = Range (int_zero, D, int_one)
+                  grid_d_steps_float = CastLike (grid_d_steps_int, step_d)
+                  grid_d_steps = Mul (grid_d_steps_float, step_d)
+                  grid_d_0 = Add (start_d, grid_d_steps)
+
+                  zeros_H_W_D = Transpose <perm = [1, 2, 0]> (zeros_D_H_W)
+                  grid_d_1 = Add (zeros_H_W_D, grid_d_0)
+                  grid_d = Transpose <perm = [2, 0, 1]>(grid_d_1)
+
+                  zeros_D_W_H = Transpose <perm = [0, 2, 1]> (zeros_D_H_W)
+                  grid_h_1 = Add (zeros_D_W_H, grid_h_0)
+                  grid_h = Transpose <perm = [0, 2, 1]>(grid_h_1)
+
+                  grid_w = Add (grid_w_0, zeros_D_H_W)
+
+                  original_grid_seq = SequenceConstruct (grid_w, grid_h, grid_d, ones_D_H_W)
+                  original_grid = ConcatFromSequence <axis: int=-1, new_axis: int=1> (original_grid_seq)
+                  constant_shape_DHW_4 = Constant <value_ints: ints = [-1, 4]> ()
+                  original_grid_DHW_4 = Reshape (original_grid, constant_shape_DHW_4)
+                  original_grid_4_DHW_ = Transpose (original_grid_DHW_4)
+
+                  original_grid_4_DHW = CastLike(original_grid_4_DHW_, theta)
+                  grid_N_3_DHW = MatMul (theta, original_grid_4_DHW)
+                  grid_N_DHW_3 = Transpose <perm = [0, 2, 1]> (grid_N_3_DHW)
+                  N_D_H_W_3_seq = SequenceConstruct (N, D, H, W, int_three_1d)
+                  N_D_H_W_3 = ConcatFromSequence <axis: int=-1, new_axis: int=0> (N_D_H_W_3_seq)
+                  grid_3d_else_ = Reshape(grid_N_DHW_3, N_D_H_W_3)
+                  grid_3d_else = CastLike(grid_3d_else_, theta)
+                  }
+              >
+        }
+        )ONNX")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          if (!hasNInputShapes(ctx, 1)) {
+            return;
+          }
+
+          const auto& size_shape = ctx.getInputType(1)->tensor_type().shape();
+          const auto size_rank = size_shape.dim_size();
+          if (size_rank != 1) {
+            fail_shape_inference("Rank of input 'size' is ", size_rank, ". It must be 1.");
+          }
+
+          const TensorProto* size_initializer = ctx.getInputData(1);
+          const auto* size_input = ctx.getSymbolicInput(1);
+          TensorShapeProto size_proto;
+          if (size_initializer) {
+            auto sizeData = ParseData<int64_t>(size_initializer);
+            for (auto val : sizeData) {
+              size_proto.add_dim()->set_dim_value(val);
+            }
+          } else if (size_input) {
+            size_proto.CopyFrom(*size_input);
+          } else {
+            return;
+          }
```

Comment:
I think you may be able to reuse the function [getInputShape](https://github.com/onnx/onnx/blob/efa11e6461960825ae5fc23ce48893ac25853bad/onnx/defs/shape_inference.h#L535) for this?

Reply:
Sorry, I attached the correct link above, but the name of the function is `getShapeInput` (which is different from `getInputShape`). This is used for any input that represents a shape. It tries to get the information from `getInputData` and `getSymbolicInput` etc. as done above). See [this implementation](https://github.com/onnx/onnx/blob/391f5703abe1a0d5c8cbb0b38df079b0bacec079/onnx/defs/shape_inference.cc#L449)

Reply:
this is very useful! I updated my change to use this function.